oak/server/rust/oak_runtime/src/node/grpc/server/mod.rs

```diff
@@ -504,64 +521,126 @@ impl GrpcRequestHandler {
+/// is always closed.
+impl Drop for GrpcResponseIterator {
+    fn drop(&mut self) {
+        debug!(
+            "Dropping GrpcResponseIterator for '{}': close channel {}",
+            self.method_name, self.response_reader
+        );
+        if let Err(err) = self.runtime.channel_close(self.response_reader) {
+            error!("Failed to close gRPC response reader channel: {:?}", err);
+        }
+        // Note that dropping self.timer will record the duration.
     }
+}
+
+impl Iterator for GrpcResponseIterator {
+    type Item = GrpcResponse;
 
-    /// Processes a gRPC response from a channel represented by `response_reader` and returns an
-    /// HTTP response body.
-    fn handle_grpc_response(&self, response_reader: oak_abi::Handle) -> Result<GrpcResponse, ()> {
+    /// Read a single encapsulated gRPC response from the provided channel.
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.done {
+            return None;
+        }
         let read_status = self
             .runtime
-            .wait_on_channels(&[response_reader])
+            .wait_on_channels(&[self.response_reader])
             .map_err(|error| {
-                error!("Couldn't wait on the temporary gRPC channel: {:?}", error);
-            })?;
-
-        let result = if read_status[0] == ChannelReadStatus::ReadReady {
-            self.runtime
-                .channel_read(response_reader)
-                .map_err(|error| {
-                    error!("Couldn't read from the temporary gRPC channel: {:?}", error);
-                })
-                .map(|message| {
-                    // Return an empty HTTP body if the `message` is None.
-                    message.map_or(vec![], |m| {
-                        self.runtime
-                            .metrics_data()
+                error!("Couldn't wait on the gRPC response channel: {:?}", error);
+            })
+            .ok()?;
+
+        if read_status[0] == ChannelReadStatus::ReadReady {
+            match self.runtime.channel_read(self.response_reader) {
+                Ok(Some(msg)) => match GrpcResponse::decode(msg.data.as_slice()) {
+                    Ok(grpc_rsp) => {
+                        self.metrics_data
                             .grpc_server_metrics
                             .grpc_response_size_bytes
                             .with_label_values(&[&self.method_name])
-                            .observe(m.data.len() as f64);
-                        m.data
-                    })
-                })
-                .and_then(|response| {
-                    // Return the serialized message body.
-                    GrpcResponse::decode(response.as_slice()).map_err(|error| {
-                        error!("Couldn't parse the GrpcResponse message: {}", error);
-                    })
-                })
+                            .observe(grpc_rsp.rsp_msg.len() as f64);
+                        if grpc_rsp.last {
```

Comment:
Will it be marked as `last` by the Oak Node that sends this stream?

Reply:
Yes if it's actually a [unary method](https://github.com/project-oak/oak/blob/master/sdk/rust/oak/src/grpc/mod.rs#L297-L300), but not necessarily if it's what turns out to be the last response for a server-streaming method.  Added a comment.